src/systems/environmental_sensor_system/EnvironmentalSensorSystem.cc

```diff
@@ -113,30 +211,87 @@ class EnvironmentalSensor : public gz::sensors::Sensor
   {
     if (!this->ready) return false;
 
-    if (!this->session.has_value()) return false;
 
-    // Step time if its not static
-    if (!this->gridField->staticTime)
-      this->session = this->gridField->frame[this->fieldName].StepTo(
-        this->session.value(), std::chrono::duration<double>(_now).count());
+    std::optional<double> dataPoints[3];
+    for (std::size_t i = 0; i < this->numberOfFields; ++i)
+    {
+      if (!this->session[i].has_value()) return false;
+
+      // Step time if its not static
+      if (!this->gridField->staticTime)
+        this->session[i] = this->gridField->frame[this->fieldName[i]].StepTo(
+          this->session[i].value(),
+          std::chrono::duration<double>(_now).count());
 
-    if (!this->session.has_value()) return false;
+      if (!this->session[i].has_value()) return false;
+
+      dataPoints[i] = this->gridField->frame[this->fieldName[i]].LookUp(
+        this->session[i].value(), this->position);
+    }
 
-    gz::msgs::Double msg;
-    *msg.mutable_header()->mutable_stamp() = gz::msgs::Convert(_now);
-    auto frame = msg.mutable_header()->add_data();
-    frame->set_key("frame_id");
-    frame->add_value((this->frameId == "") ? this->Name() : this->frameId);
-    auto data = this->gridField->frame[this->fieldName].LookUp(
-      this->session.value(), this->position);
-    if (!data.has_value())
+    if(this->numberOfFields == 1) {
+      gz::msgs::Double msg;
+      *msg.mutable_header()->mutable_stamp() = gz::msgs::Convert(_now);
+      auto frame = msg.mutable_header()->add_data();
+      frame->set_key("frame_id");
+      frame->add_value((this->frameId == "") ? this->Name() : this->frameId);
+      auto data = dataPoints[0];
+      if (!data.has_value())
+      {
+        gzwarn << "Failed to acquire value perhaps out of field?\n";
+        return false;
+      }
+      msg.set_data(data.value());
+      // TODO(anyone) Add sensor noise.
+      this->pub.Publish(msg);
+    }
+    else if (this->numberOfFields == 2)
     {
-      gzwarn << "Failed to acquire value perhaps out of field?\n";
-      return false;
+      gz::msgs::Vector2d msg;
+      *msg.mutable_header()->mutable_stamp() = gz::msgs::Convert(_now);
+      auto frame = msg.mutable_header()->add_data();
+      frame->set_key("frame_id");
+      frame->add_value((this->frameId == "") ? this->Name() : this->frameId);
+
+      if (!dataPoints[0].has_value() || !dataPoints[1].has_value())
+      {
+        gzwarn << "Failed to acquire value perhaps out of field?\n";
+        return false;
+      }
+
+      math::Vector3d vector(dataPoints[0].value(), dataPoints[1].value(), 0);
+      auto transformed = transformFrame(this->transformType, this->objectPose,
```

Comment:
@arjo129 meta: hmm, a 3D transform for 2D data? On z = 0? Why z = 0?

Reply:
Thinking out loud, Gazebo performs rigid body simulation in 3D. Does it make sense to have 2D data? Or is 2D data actually 3D data in some plane (or all planes)?

Reply:
I'm not too sure how we should handle a 2D case. Our current example format for MBARI uses such a 2D case which is why I default to zero.

Reply:
> Perhaps if we see a z component is missing we assume that the thing defaults to zero.

That sounds pretty reasonable.

Reply:
4b2ea71 does this.